File: src/transformers/ImportTransformer.ts
import path from "path";
import ts from "typescript";

export namespace ImportTransformer {
  export const transform =
    (props: { from: string; to: string }) =>
    (context: ts.TransformationContext) =>
    (file: ts.SourceFile) =>
      transform_file({
        top: props.from,
        to: props.to,
        context,
        file,
      });

  const transform_file = (props: {
    top: string;
    to: string;
    context: ts.TransformationContext;
    file: ts.SourceFile;
  }): ts.SourceFile => {
    if (props.file.isDeclarationFile) return props.file;

    const from: string = get_directory_path(
      path.resolve(props.file.getSourceFile().fileName),
    );
    const to: string = from.replace(props.top, props.to);

    // First pass: transform relative imports
    let transformedFile = ts.visitEachChild(
      props.file,
      (node) =>
        transform_node({
          top: props.top,
          from,
          to,
          node,
        }),
      props.context,
    );

    // Second pass: remove unused typia imports
    transformedFile = removeUnusedTypiaImports(transformedFile);

    return transformedFile;
  };

  const transform_node = (props: {
    top: string;
    from: string;
    to: string;
    node: ts.Node;
  }) => {
    if (
      !ts.isImportDeclaration(props.node) ||
      !ts.isStringLiteral(props.node.moduleSpecifier)
    )
      return props.node;

    const text: string = props.node.moduleSpecifier.text;
    if (text[0] !== ".") return props.node;

    const location: string = path.resolve(props.from, text);
    if (location.indexOf(props.top) === 0) return props.node;

    const replaced: string = (() => {
      const simple: string = path
        .relative(props.to, location)
        .split(path.sep)
        .join("/");
      return simple[0] === "." ? simple : `./${simple}`;
    })();

    return ts.factory.createImportDeclaration(
      undefined,
      props.node.importClause,
      ts.factory.createStringLiteral(replaced),
      props.node.assertClause,
    );
  };
}

const get_directory_path = (file: string): string => {
  const split: string[] = path.resolve(file).split(path.sep);
  split.pop();
  return path.resolve(split.join(path.sep));
};

/** Remove unused typia imports that are only used in transformable calls */
const removeUnusedTypiaImports = (file: ts.SourceFile): ts.SourceFile => {
  // Find typia imports and collect all identifiers
  interface ImportMetadata {
    declaration: ts.ImportDeclaration;
    default: boolean;
  }
  const imports: Map<string, ImportMetadata> = new Map();
  for (const stmt of file.statements) {
    if (
      ts.isImportDeclaration(stmt) === false ||
      ts.isStringLiteral(stmt.moduleSpecifier) === false ||
      stmt.moduleSpecifier.text !== "typia" ||
      stmt.importClause === undefined
    )
      continue;

    // Track default import (import typia from 'typia')
    if (stmt.importClause.name)
      imports.set(stmt.importClause.name.text, {
        declaration: stmt,
        default: true,
      });

    // Track named imports (import { tags } from 'typia') - keep these
    if (
      stmt.importClause.namedBindings &&
      ts.isNamedImports(stmt.importClause.namedBindings)
    )
      for (const element of stmt.importClause.namedBindings.elements)
        imports.set(element.name.text, {
          declaration: stmt,
          default: false,
        });
  }
  if (imports.size === 0) return file; // No typia imports to check

  // Find usage of typia identifiers that are NOT transformable calls
  const nonTransformableUsage = new Set<string>();
  const checkUsage = (node: ts.Node) => {
    if (ts.isIdentifier(node) && imports.has(node.text)) {
      const identifier: string = node.text;
      // Check if this identifier is being used as part of a property access
      if (
        node.parent &&
        ts.isPropertyAccessExpression(node.parent) &&
        node.parent.expression === node
      ) {
        // This is typia.something - check if it's a transformable call pattern
        if (!isLikelyTransformableCall(node.parent))
          nonTransformableUsage.add(identifier);
      } else
        // Direct usage of the typia identifier (not as property access)
        // This is definitely non-transformable usage
        nonTransformableUsage.add(identifier);
    }
    ts.forEachChild(node, checkUsage);
  };

  // Check all statements except import declarations
  for (const stmt of file.statements)
    if (
      !ts.isImportDeclaration(stmt) ||
      !ts.isStringLiteral(stmt.moduleSpecifier) ||
      stmt.moduleSpecifier.text !== "typia"
    )
      checkUsage(stmt);

  // Update import statements
  const newStatements: ts.Statement[] = file.statements
    .map((stmt) => {
      if (
        ts.isImportDeclaration(stmt) &&
        ts.isStringLiteral(stmt.moduleSpecifier) &&
        stmt.moduleSpecifier.text === "typia" &&
        stmt.importClause
      ) {
        const newImportClause = filterTypiaImportClause(
          stmt.importClause,
          nonTransformableUsage,
        );
        if (newImportClause)
          return ts.factory.createImportDeclaration(
            stmt.modifiers,
            newImportClause,
            stmt.moduleSpecifier,
            stmt.attributes,
          );
        return null; // Skip adding the import if all imports are unused
      }
      return stmt;
    })
    .filter((stmt) => stmt !== null);
  return ts.factory.updateSourceFile(
    file,
    newStatements,
    file.isDeclarationFile,
    file.referencedFiles,
    file.typeReferenceDirectives,
    file.hasNoDefaultLib,
    file.libReferenceDirectives,
  );
};

/**
 * Check if a property access expression looks like a transformable typia call
 * This uses heuristics to detect patterns like typia.xxx(),
 * typia.namespace.xxx()
 */
const isLikelyTransformableCall = (
  node: ts.PropertyAccessExpression,
): boolean => {
  // Check if this is eventually part of a call expression
  let current: ts.Node = node;

  // Walk up the chain to find if this leads to a call expression
  // Handle patterns like: typia.xxx(), typia.namespace.xxx()
  while (ts.isPropertyAccessExpression(current)) {
    current = current.parent;
  }

  // If the final result is a call expression, this is likely transformable
  if (
    ts.isCallExpression(current) &&
    (current.expression === node ||
      (ts.isPropertyAccessExpression(current.expression) &&
        isTypiaPropertyChain(current.expression)))
  ) {
    return true;
  }

  return false;
};

/** Check if a property access expression is part of a typia.xxx.yyy chain */
const isTypiaPropertyChain = (node: ts.PropertyAccessExpression): boolean => {
  let current: ts.Expression = node;

  while (ts.isPropertyAccessExpression(current)) {
    current = current.expression;
  }

  return ts.isIdentifier(current) && current.text === "typia";
};

/** Filter import clause to remove unused default imports */
const filterTypiaImportClause = (
  importClause: ts.ImportClause,
  usedImports: Set<string>,
): ts.ImportClause | undefined => {
  const hasDefaultImport =
    importClause.name && usedImports.has(importClause.name.text);
  const namedBindings = importClause.namedBindings; // Always keep named bindings like { tags }

  // Return undefined if no imports are used
  if (!hasDefaultImport && !namedBindings) {
    return undefined;
  }

  return ts.factory.createImportClause(
    importClause.isTypeOnly,
    hasDefaultImport ? importClause.name : undefined,
    namedBindings,
  );
};

File: packages/typia/native/transform/ImportTransformer.go
package transform

import (
	"path/filepath"
	"strings"

	shimast "github.com/microsoft/typescript-go/shim/ast"
)

type importTransformerNamespace struct{}

var ImportTransformer = importTransformerNamespace{}

type ImportTransformer_TransformProps struct {
	From string
	To   string
}

type ImportTransformer_Type func(file *shimast.SourceFile) *shimast.SourceFile

type importTransformer_transformFileProps struct {
	Top     string
	To      string
	Context any
	File    *shimast.SourceFile
}

type importTransformer_transformNodeProps struct {
	Top  string
	From string
	To   string
	Node *shimast.Node
}

type importTransformer_importMetadata struct {
	Declaration *shimast.ImportDeclaration
	Default     bool
}

var importTransformer_factory = shimast.NewNodeFactory(shimast.NodeFactoryHooks{})

func (importTransformerNamespace) Transform(props ImportTransformer_TransformProps) func(context any) ImportTransformer_Type {
	return func(context any) ImportTransformer_Type {
		return func(file *shimast.SourceFile) *shimast.SourceFile {
			return importTransformer_transform_file(importTransformer_transformFileProps{
				Top:     props.From,
				To:      props.To,
				Context: context,
				File:    file,
			})
		}
	}
}

func importTransformer_transform_file(props importTransformer_transformFileProps) *shimast.SourceFile {
	if props.File == nil || props.File.IsDeclarationFile {
		return props.File
	}
	from := get_directory_path(filepath.Clean(props.File.FileName()))
	to := strings.Replace(from, filepath.Clean(props.Top), filepath.Clean(props.To), 1)

	var visitor *shimast.NodeVisitor
	visitor = shimast.NewNodeVisitor(func(node *shimast.Node) *shimast.Node {
		if node == nil {
			return nil
		}
		next := importTransformer_transform_node(importTransformer_transformNodeProps{
			Top:  props.Top,
			From: from,
			To:   to,
			Node: node,
		})
		return visitor.VisitEachChild(next)
	}, importTransformer_factory, shimast.NodeVisitorHooks{})

	transformed := visitor.VisitNode(props.File.AsNode())
	if transformed == nil {
		return props.File
	}
	return removeUnusedTypiaImports(transformed.AsSourceFile())
}

func importTransformer_transform_node(props importTransformer_transformNodeProps) *shimast.Node {
	if props.Node == nil || props.Node.Kind != shimast.KindImportDeclaration {
		return props.Node
	}
	decl := props.Node.AsImportDeclaration()
	if decl == nil || decl.ModuleSpecifier == nil || decl.ModuleSpecifier.Kind != shimast.KindStringLiteral {
		return props.Node
	}
	text := decl.ModuleSpecifier.AsStringLiteral().Text
	if text == "" || text[0] != '.' {
		return props.Node
	}

	location := filepath.Clean(filepath.Join(props.From, text))
	top := filepath.Clean(props.Top)
	if strings.HasPrefix(location, top) {
		return props.Node
	}

	replaced, err := filepath.Rel(props.To, location)
	if err != nil {
		return props.Node
	}
	replaced = filepath.ToSlash(replaced)
	if strings.HasPrefix(replaced, ".") == false {
		replaced = "./" + replaced
	}
	return importTransformer_factory.UpdateImportDeclaration(
		decl,
		decl.Modifiers(),
		decl.ImportClause,
		importTransformer_factory.NewStringLiteral(replaced, shimast.TokenFlagsNone),
		decl.Attributes,
	)
}

func get_directory_path(file string) string {
	return filepath.Clean(filepath.Dir(file))
}

func removeUnusedTypiaImports(file *shimast.SourceFile) *shimast.SourceFile {
	if file == nil || file.Statements == nil {
		return file
	}
	imports := map[string]importTransformer_importMetadata{}
	for _, stmt := range file.Statements.Nodes {
		decl := importTransformer_typiaImport(stmt)
		if decl == nil || decl.ImportClause == nil {
			continue
		}
		clause := decl.ImportClause.AsImportClause()
		if clause == nil {
			continue
		}
		if name := importTransformer_identifierText(clause.Name()); name != "" {
			imports[name] = importTransformer_importMetadata{Declaration: decl, Default: true}
		}
		if clause.NamedBindings != nil && clause.NamedBindings.Kind == shimast.KindNamedImports {
			named := clause.NamedBindings.AsNamedImports()
			if named != nil && named.Elements != nil {
				for _, elem := range named.Elements.Nodes {
					if elem == nil {
						continue
					}
					spec := elem.AsImportSpecifier()
					if spec == nil {
						continue
					}
					if name := importTransformer_identifierText(spec.Name()); name != "" {
						imports[name] = importTransformer_importMetadata{Declaration: decl, Default: false}
					}
				}
			}
		}
	}
	if len(imports) == 0 {
		return file
	}

	nonTransformableUsage := map[string]bool{}
	var checkUsage func(node *shimast.Node)
	checkUsage = func(node *shimast.Node) {
		if node == nil {
			return
		}
		if node.Kind == shimast.KindIdentifier {
			identifier := node.AsIdentifier().Text
			if _, ok := imports[identifier]; ok {
				parent := node.Parent
				if parent != nil && parent.Kind == shimast.KindPropertyAccessExpression {
					access := parent.AsPropertyAccessExpression()
					if access != nil && access.Expression == node {
						if importTransformer_isLikelyTransformableCall(parent) == false {
							nonTransformableUsage[identifier] = true
						}
					}
				} else {
					nonTransformableUsage[identifier] = true
				}
			}
		}
		node.ForEachChild(func(child *shimast.Node) bool {
			checkUsage(child)
			return false
		})
	}
	for _, stmt := range file.Statements.Nodes {
		if importTransformer_typiaImport(stmt) == nil {
			checkUsage(stmt)
		}
	}

	nodes := make([]*shimast.Node, 0, len(file.Statements.Nodes))
	for _, stmt := range file.Statements.Nodes {
		decl := importTransformer_typiaImport(stmt)
		if decl == nil || decl.ImportClause == nil {
			nodes = append(nodes, stmt)
			continue
		}
		clause := decl.ImportClause.AsImportClause()
		if clause == nil {
			nodes = append(nodes, stmt)
			continue
		}
		name := clause.Name()
		if importTransformer_identifierText(name) != "" && nonTransformableUsage[importTransformer_identifierText(name)] == false {
			name = nil
		}
		namedBindings := clause.NamedBindings
		if name == nil && namedBindings == nil {
			continue
		}
		nextImportClause := importTransformer_factory.NewImportClause(
			clause.PhaseModifier,
			name,
			namedBindings,
		).AsImportClause()
		nodes = append(nodes, importTransformer_factory.UpdateImportDeclaration(
			decl,
			decl.Modifiers(),
			nextImportClause.AsNode(),
			decl.ModuleSpecifier,
			decl.Attributes,
		))
	}
	return importTransformer_factory.UpdateSourceFile(
		file,
		importTransformer_factory.NewNodeList(nodes),
		file.EndOfFileToken,
	).AsSourceFile()
}

func importTransformer_typiaImport(stmt *shimast.Node) *shimast.ImportDeclaration {
	if stmt == nil || stmt.Kind != shimast.KindImportDeclaration {
		return nil
	}
	decl := stmt.AsImportDeclaration()
	if decl == nil || decl.ModuleSpecifier == nil || decl.ModuleSpecifier.Kind != shimast.KindStringLiteral {
		return nil
	}
	if decl.ModuleSpecifier.AsStringLiteral().Text != "typia" {
		return nil
	}
	return decl
}

func importTransformer_isLikelyTransformableCall(node *shimast.Node) bool {
	current := node
	for current != nil && current.Kind == shimast.KindPropertyAccessExpression {
		current = current.Parent
	}
	if current == nil || current.Kind != shimast.KindCallExpression {
		return false
	}
	call := current.AsCallExpression()
	if call == nil {
		return false
	}
	if call.Expression == node {
		return true
	}
	return call.Expression != nil &&
		call.Expression.Kind == shimast.KindPropertyAccessExpression &&
		importTransformer_isTypiaPropertyChain(call.Expression)
}

func importTransformer_isTypiaPropertyChain(node *shimast.Node) bool {
	current := node
	for current != nil && current.Kind == shimast.KindPropertyAccessExpression {
		current = current.AsPropertyAccessExpression().Expression
	}
	return importTransformer_identifierText(current) == "typia"
}

func importTransformer_identifierText(node *shimast.Node) string {
	if node == nil || node.Kind != shimast.KindIdentifier {
		return ""
	}
	return node.AsIdentifier().Text
}
